lstchain/scripts/lstchain_r0_to_r0g.py

```diff
@@ -0,0 +1,206 @@
+
+                if ((evtype == EventType.SUBARRAY.value) |
+                    (evtype == EventType.UNKNOWN.value)):
+                    # Find pixels with HG above gain switch threshold:
+                    wf = protozfits.any_array_to_numpy(event.waveform)
+                    num_gains = int(wf.size / num_pixels / num_samples)
+
+                    if num_gains != 2:
+                        logging.error('You are attempting gain selection on '
+                                      'data with only one gain!')
+                        exit(1)
+
+                    wf = wf.reshape((num_gains, num_pixels, num_samples))
+                    hg = wf[0]
+                    lg = wf[1]
+                    use_lg = (np.max(hg[:, SAMPLE_START:SAMPLE_END], axis=1)
+                              > THRESHOLD)
+                    new_wf = np.where(use_lg[:, None], lg, hg) # gain-selected
+                    event.waveform.data = new_wf.tobytes()
+                    event.num_channels = 1  # Just one gain stored!
+
+                    pixel_status = protozfits.any_array_to_numpy(event.pixel_status)
+                    # Set to 0 the status bit of the removed gain:
+                    new_status = np.where(use_lg,
+                                          pixel_status & 0b1011,
+                                          pixel_status & 0b0111)
+                    event.pixel_status.data = new_status.tobytes()
+
+                stream.write_message(event)
+
+            stream.close()
+            input_streams[i].close()
+
+    logging.info('R0 to R0G conversion finished successfully!')
+    return(0)
+
+
+def get_event_types(input_file):
+
+    # For heuristic flat field identification (values refer to
+    # baseline-subtracted HG integrated waveforms):
+    MIN_FLATFIELD_ADC = 3000
+    MAX_FLATFIELD_ADC = 12000
+    MIN_FLATFIELD_PIXEL_FRACTION = 0.8
+
+    standard_config['source_config']['LSTEventSource'][
+        'apply_drs4_corrections'] = False
+    standard_config['source_config']['LSTEventSource'][
+        'pointing_information'] = False
+
+    event_type = []
+    event_id = []
+    with EventSource(input_url=input_file,
+                     config=Config(standard_config['source_config'])) as source:
+        source.pointing_information = False
+        source.trigger_information = True
+        source.log.setLevel(logging.WARNING)
+        try:
+            for event in source:
+                if event.r0.tel[1].waveform is None:
+                    logging.error('The data seem to contain no R0 waveforms. '
+                                  'Is this already gain-selected data?')
+                    exit(1)
+
+                # Check if this may be a FF event
+                # Subtract baseline offset (convert to signed integer to
+                # allow for negative fluctuations!)
+                wf_hg = event.r0.tel[1].waveform[0][:,
+                        SAMPLE_START:SAMPLE_END].astype('int16') - OFFSET
```

Comment:
Subtracting the offset is wrong for EVBv6 data. It will be done automatically by the source since the offset is now stored in the file.

Reply:
> Subtracting the offset is wrong for EVBv6 data. It will be done automatically by the source since the offset is now stored in the file.

Is that also the case if all DRS4 corrections are disabled? Can't we read with the event source the waveforms as they are stored?

Reply:
> > Subtracting the offset is wrong for EVBv6 data. It will be done automatically by the source since the offset is now stored in the file.
> 
> Is that also the case if all DRS4 corrections are disabled? Can't we read with the event source the waveforms as they are stored?

I confirm I had tried to use the FF heuristic identification in the event source, but it only works if the calibration is switched on. And of course, we do not have it at the point at which we want to make the gain selection.

Reply:
> Is that also the case if all DRS4 corrections are disabled? 

Yes, it's the first step when reading the waveforms. This is completely independent of DRS4 corrections.

The scaled waveform is put into `r1`, if no pe calibration has been performed, the unscaled waveform is also stored, in r0.

But I don't understand why you would want to reimplement this here instead of just using what we already have.

This is now a manual translation of the C++ code back to python which was a manual translation of the code in ctapipe_io_lst to C++,